web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TraceContextCorrelation.java

```diff
@@ -0,0 +1,448 @@
+package com.microsoft.applicationinsights.web.internal.correlation;
+
+import com.google.common.base.Joiner;
+import com.microsoft.applicationinsights.TelemetryConfiguration;
+import com.microsoft.applicationinsights.internal.logger.InternalLogger;
+import com.microsoft.applicationinsights.telemetry.RequestTelemetry;
+import com.microsoft.applicationinsights.web.internal.RequestTelemetryContext;
+import com.microsoft.applicationinsights.web.internal.ThreadContext;
+import com.microsoft.applicationinsights.web.internal.correlation.tracecontext.Traceparent;
+import com.microsoft.applicationinsights.web.internal.correlation.tracecontext.Tracestate;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+
+/**
+ * A class that is responsible for performing correlation based on W3C protocol.
+ * This is a clean implementation of W3C protocol and doesn't have the backward
+ * compatibility with AI-RequestId protocol.
+ *
+ * @author Dhaval Doshi
+ */
+public class TraceContextCorrelation {
+
+    public static final String TRACEPARENT_HEADER_NAME = "traceparent";
+    public static final String TRACESTATE_HEADER_NAME = "tracestate";
+    public static final String REQUEST_CONTEXT_HEADER_NAME = "Request-Context";
+    public static final String AZURE_TRACEPARENT_COMPONENT_INITIAL = "az";
+    public static final String REQUEST_CONTEXT_HEADER_APPID_KEY = "appId";
+
+    /**
+     * Private constructor as we don't expect to create an object of this class.
+     */
+    private TraceContextCorrelation() {}
+
+    /**
+     * This method is responsible to perform correlation for incoming request by populating it's
+     * traceId, spanId and parentId. It also stores incoming tracestate into ThreadLocal for downstream
+     * propagation.
+     * @param request
+     * @param response
+     * @param requestTelemetry
+     */
+    public static void resolveCorrelation(HttpServletRequest request, HttpServletResponse response,
+        RequestTelemetry requestTelemetry) {
+
+        try {
+            if (request == null) {
+                InternalLogger.INSTANCE.error("Failed to resolve correlation. request is null.");
+                return;
+            }
+
+            if (response == null) {
+                InternalLogger.INSTANCE.error("Failed to resolve correlation. response is null.");
+                return;
+            }
+
+            if (requestTelemetry == null) {
+                InternalLogger.INSTANCE.error("Failed to resolve correlation. requestTelemetry is null.");
+                return;
+            }
+
+            Traceparent incomingTraceparent = processIncomingTraceparent(request);
+            Traceparent outGoingTraceParent = createOutgoingTraceparent(incomingTraceparent);
```

Comment:
nit. this is not outgoing request thing. It describes incoming request processing. And incomingTraceparent is actually parent thing.

Reply:
Sure thing! Will update.

Reply:
Done